containers/daap-recreate-data-product/src/var/task/recreate_data_product.py

```diff
@@ -0,0 +1,69 @@
+import os
+
+import boto3
+
+
+def handler(event, context):
+    data_product_to_recreate = event.get("data_product", "")
+    raw_data_bucket = os.environ.get("RAW_DATA_BUCKET", "")
+    curated_data_bucket = os.environ.get("CURATED_DATA_BUCKET", "")
+    athena_load_lambda = os.environ.get("ATHENA_LOAD_LAMBDA", "")
+
+    # Check data product has associated data
+    s3 = boto3.client("s3")
+    s3_response = s3.list_objects_v2(
+        Bucket=raw_data_bucket, Prefix=f"raw_data/{data_product_to_recreate}"
+    )
+    data_product_registration = s3_response.get("Contents", [])
+    if not any(data_product_registration):
+        raise ValueError(f"No data product found for {data_product_to_recreate}")
+    print(data_product_registration)
+
+    # Drop existing athena tables for that data product
+    glue = boto3.client("glue")
+    glue_response = glue.get_tables(DatabaseName=data_product_to_recreate)
+    data_product_tables = glue_response.get("TableList", [])
+    if not any(data_product_tables):
+        print(f"No tables found for data product {data_product_to_recreate}")
+    for table in data_product_tables:
+        glue.delete_table(DatabaseName=data_product_to_recreate, Name=table["Name"])
+
+    # Remove curated data files for that data product
+    s3_recursive_delete(
+        bucket=curated_data_bucket,
+        prefix=f"curated_data/database_name={data_product_to_recreate}/",
+        s3_client=s3,
+    )
+
+    # Feed all data files through the load process again. Curated files are recreated.
+    # If there are over 1000 files, the lambda will get jammed up
+    aws_lambda = boto3.client("lambda")
+    for file in data_product_registration:
+        key = file["Key"]
+        payload = f'{{"detail":{{"bucket":{{"name":"{raw_data_bucket}"}}, "object":{{"key":"{key}"}}}}}}'
+        aws_lambda.invoke(
```

Comment:
Can you introduce threading. If we going to recreate a data product there could be files more than 1000 

eg : Pseudo code below 
```
# Number of invocations
num_invocations = 2000

# Create threads and start invoking Lambda
threads = []
for _ in range(num_invocations):
    thread = threading.Thread(target=invoke_lambda)
    threads.append(thread)
    thread.start()

# Wait for all threads to finish
for thread in threads:
    thread.join()
```

Reply:
the API call here doesn't report on completion but I added a waiter every 1000 calls to it

Reply:
The lambda is invoking athena lambda ( which is responsible for processing the file . The processing time may vary. i am not sure if  **# Wait 5 mins for other invocations to finish
            time.sleep(360)** will work. It may take longer time finish processing. 

Since we are going to call this **Athena_load_lambda**  few times its better to send response back, so you can use threading 

If this means extra work in another lambda you can very well split it into another ticket.